README.md
# ExPesel

**Elixir library for PESEL number.**

  * `ExPesel.valid?/1` - you can check if PESEL number is valid
  * `ExPesel.valid_with?/2` - you can check if PESEL number is valid with additional check about sex or date of birth
  * `ExPesel.birthdate/1` - you can obtain date of birth from PESEL number
  * `ExPesel.sex/1` - you can obtain sex from the PESEL number
  * `ExPesel.zombie?/1` - you can check is PESEL number is a zombie (more than 123 years before today)

  More about PESEL number:
  * [Wikipedia EN](https://en.wikipedia.org/wiki/PESEL)
  * [Wikipedia PL](https://pl.wikipedia.org/wiki/PESEL)
  * [obywatel.gov.pl](https://obywatel.gov.pl/dokumenty-i-dane-osobowe/czym-jest-numer-pesel)

## Usage

```elixir
iex> ExPesel.valid?("44051401458")
true

iex> ExPesel.valid?("90720312611")
false

iex> ExPesel.valid_with?("44051401458", :male)
true

iex> ExPesel.valid_with?("88122302080", :male)
false

iex> ExPesel.valid_with?("44051401458", :female)
false

iex> ExPesel.valid_with?("88122302080", :female)
true

iex> ExPesel.valid_with?("44051401458", {1944, 5, 14})
true

iex> ExPesel.valid_with?("44051401458", {1944, 6, 13})
false

iex> ExPesel.birthdate("10320305853")
{2010, 12, 3}

iex> ExPesel.sex("88122302080")
:female

iex> ExPesel.zombie?("01920300359")
true
```

## Installation

Package can be installed by adding `ex_pesel`
to your list of dependencies in `mix.exs`:

```elixir
def deps do
  [
    {:ex_pesel, "~> 0.2.0"}
  ]
end
```

## Documentation

Docs can be found at [https://hexdocs.pm/ex_pesel](https://hexdocs.pm/ex_pesel).

## License

Source code of ex_pesel is released under the Apache 2.0 license, see the [LICENSE](LICENSE) file.


CHANGELOG.md
# Changelog

## v0.2.0

 * fix ArgumentError when passing broken PESEL containing not only digits

## v0.1.0

 * here it begins


test/ex_pesel_test.exs
defmodule ExPeselTest do
  use ExUnit.Case
  doctest ExPesel

  test "PESEL validation only works for string not integer" do
    assert catch_error(ExPesel.valid?(12345678901)) == :function_clause
  end

  test "zombie PESEL is more than 123 years before than today: 2018-01-07" do
    # TODO:
  end
end


test/ex_pesel/pesel_test.exs
defmodule ExPesel.PeselTest do
  use ExUnit.Case
  doctest ExPesel.Pesel
end


mix.exs
defmodule ExPesel.MixProject do
  use Mix.Project

  @version "0.2.0"

  @description "Library for PESEL number."
  @repo_url "https://github.com/crabonature/ex_pesel"
  @docs "https://hexdocs.pm/ex_pesel"

  def project do
    [
      app: :ex_pesel,
      version: @version,
      elixir: "~> 1.5",
      start_permanent: Mix.env() == :prod,
      deps: deps(),

      # Hex
      description: @description,
      package: package(),
      # docs
      homepage_url: @docs,
      source_url: @repo_url
    ]
  end

  # Run "mix help compile.app" to learn about applications.
  def application do
    [
      extra_applications: [:logger]
    ]
  end

  defp package do
    [
      files: ["lib/**/*.ex", "mix.exs", "README.md", "LICENSE"],
      maintainers: ["Marcin Szczepaniak"],
      licenses: ["Apache 2.0"],
      links: %{
	"GitHub" => @repo_url,
	"Docs" => @docs
      }
    ]
  end

  # Run "mix help deps" to learn about dependencies.
  defp deps do
    [
      {:ex_doc,   "~> 0.18.1", only: :dev, runtime: false},
      {:dialyxir, "~> 0.5",    only: :dev, runtime: false},
      {:credo,    "~> 0.8",    only: [:dev, :test], runtime: false}
    ]
  end
end


config/config.exs
# This file is responsible for configuring your application
# and its dependencies with the aid of the Mix.Config module.
use Mix.Config

# This configuration is loaded before any dependency and is restricted
# to this project. If another project depends on this project, this
# file won't be loaded nor affect the parent project. For this reason,
# if you want to provide default values for your application for
# 3rd-party users, it should be done in your "mix.exs" file.

# You can configure your application as:
#
#     config :ex_pesel, key: :value
#
# and access this configuration in your application as:
#
#     Application.get_env(:ex_pesel, :key)
#
# You can also configure a 3rd-party app:
#
#     config :logger, level: :info
#

# It is also possible to import configuration files, relative to this
# directory. For example, you can emulate configuration per environment
# by uncommenting the line below and defining dev.exs, test.exs and such.
# Configuration from the imported file will override the ones defined
# here (which is why it is important to import them last).
#
#     import_config "#{Mix.env}.exs"


lib/ex_pesel/pesel.ex
defmodule ExPesel.Pesel do
  @moduledoc false

  # wages for calculating PESEL checksum
  @w [9, 7, 3, 1, 9, 7, 3, 1, 9, 7]

  # longest human lifespan in years
  @years 123

  # PESEL is valid
  # when checksum for 10 first digits is equal to 11 digit
  # when first 6 digits gives proper date
  # when first 6 digits gives date until today, PESEL for future is invalid
  @spec valid?(String.t()) :: boolean()
  def valid?(pesel) when is_binary(pesel) do
    checksum?(to_list(pesel), @w, 0)
    and valid_birthdate?(birthdate(pesel))
  end

  defp checksum?([crc], [], acc),         do: crc == rem(acc, 10)
  defp checksum?([x | r], [w | rw], acc), do: checksum?(r, rw, x * w + acc)
  defp checksum?(_, _, _),                do: false

  defp valid_birthdate?(date),  do: valid_date?(date) and till_today?(date)
  defp valid_date?(date),       do: :calendar.valid_date(date)
  defp till_today?(date),       do: date <= today()
  defp today,                   do: extract_date(:calendar.universal_time)
  defp extract_date({date, _}), do: date

  # PESEL is valid and was issued for specified birthdate
  @spec valid_with?(String.t(), {1800..2299, 1..12, 1..31}) :: boolean()
  def valid_with?(pesel, birthday)
  when is_binary(pesel) and is_tuple(birthday) do
    valid?(pesel) and birthdate(pesel) == birthday
  end

  # PESEL is valid and was issued for specified sex
  @spec valid_with?(String.t(), :male|:female) :: boolean()
  def valid_with?(pesel, sex)
  when is_binary(pesel) and is_atom(sex) do
    valid?(pesel) and sex(pesel) == sex
  end

  # Date of birth is encoded in first 6 digits as yymmdd
  # in month mm there is encoded century as:
  # for 1800-1899 we got addition of 80 to mm
  # for 1900-1999 we got addition of  0 to mm
  # for 2000-2099 we got addition of 20 to mm
  # for 2100-2199 we got addition of 40 to mm
  # for 2200-2299 we got addition of 60 to mm
  @spec birthdate(String.t()) :: {1800..2299, 1..12, 1..31} | :unknown
  def birthdate(pesel), do: do_birthdate(to_list(pesel))

  # sorry for that but in elixir there is no multiguards
  defp do_birthdate([y1, y2, m1, m2, d1, d2 | _])
  when m1 < 2, do: date({19, y1, y2}, {m1, m2}, {d1, d2})

  defp do_birthdate([y1, y2, m1, m2, d1, d2 | _])
  when m1 < 4, do: date({20, y1, y2}, {m1 - 2, m2}, {d1, d2})

  defp do_birthdate([y1, y2, m1, m2, d1, d2 | _])
  when m1 < 6, do: date({21, y1, y2}, {m1 - 4, m2}, {d1, d2})

  defp do_birthdate([y1, y2, m1, m2, d1, d2 | _])
  when m1 < 8, do: date({22, y1, y2}, {m1 - 6, m2}, {d1, d2})

  defp do_birthdate([y1, y2, m1, m2, d1, d2 | _])
  when m1 < 10, do: date({18, y1, y2}, {m1 - 8, m2}, {d1, d2})

  defp do_birthdate([]), do: :unknown

  defp date(as_year, as_month, as_day) do
    {year(as_year), month(as_month), day(as_day)}
  end

  defp day({d1, d2}),           do: d1 * 10 + d2
  defp month({m1, m2}),         do: m1 * 10 + m2
  defp year({century, y1, y2}), do: century * 100 + y1 * 10 + y2

  # Sex is encoded with 10 digit
  # odd for male
  # even for female
  @spec sex(String.t()) :: :male | :female | :unknown
  def sex(pesel) when is_binary(pesel), do: do_sex(to_list(pesel))

  defp do_sex([_, _, _, _, _, _, _, _, _, s | _]) when rem(s, 2) == 1, do: :male
  defp do_sex([]), do: :unknown
  defp do_sex(_),  do: :female

  # PESEL is zombie if
  # date of birth is 123 years before today
  @spec zombie?(String.t()) :: boolean()
  def zombie?(pesel) when is_binary(pesel), do: is_zombie?(pesel)

  defp is_zombie?(pesel), do: dead?(birthdate(pesel))
  defp dead?(:unknown), do: false
  defp dead?(birthday), do: birthday < years_before(today(), @years)
  defp years_before({year, month, day}, dist), do: {year - dist, month, day}

  # Transpose from string to list
  defp to_list(pesel) do
    pesel
    |> scan()
    |> resolve()
    |> String.split("", trim: true)
    |> Enum.map(&String.to_integer/1)
  end

  defp scan(pesel), do: Regex.scan(~r/^\d{11}$/, pesel)
  defp resolve([]),         do: ""
  defp resolve([[pesel]]),  do: pesel
end


lib/ex_pesel.ex
defmodule ExPesel do
  @moduledoc """
  Library for PESEL number.

  * `ExPesel.valid?/1` - you can check if PESEL number is valid
  * `ExPesel.valid_with?/2` - you can check if PESEL number is valid with additional check about sex or birthdate
  * `ExPesel.birthdate/1` - you can obtain date of birth from PESEL number
  * `ExPesel.sex/1` - you can obtain sex from the PESEL number
  * `ExPesel.zombie?/1` - you can obtain is PESEL number is a zombie (more than 123 years before today)

  More about PESEL number:
  * [Wikipedia EN](https://en.wikipedia.org/wiki/PESEL)
  * [Wikipedia PL](https://pl.wikipedia.org/wiki/PESEL)
  * [obywatel.gov.pl](https://obywatel.gov.pl/dokumenty-i-dane-osobowe/czym-jest-numer-pesel)
  """

  alias ExPesel.Pesel

  @doc """
  Is PESEL number valid?

  PESEL is valid when:
    * length of it is 11
    * last digit is proper checksum for first ten digits
    * date of birt from first 6 digits are proper
    * date of birth is until today.

  For example:
      iex> ExPesel.valid?("44051401458")
      true

      iex> ExPesel.valid?("90720312611")
      false

      iex> ExPesel.valid?("90520308014")
      false

      iex> ExPesel.valid?("44051401459")
      false

      iex> ExPesel.valid?("00000000000")
      false

      iex> ExPesel.valid?("234555")
      false

      iex> ExPesel.valid?("23455532312131123123")
      false

      iex> ExPesel.valid?("some really bad data")
      false
  """
  @spec valid?(String.t()) :: boolean()
  defdelegate valid?(pesel), to: Pesel

  @doc """
  Is PESEL valid and additionaly:
  * belongs to male?
  * or belongs to female?
  * or date of birth extracted from PESEL is equal birthdate?

  For example:

      iex> ExPesel.valid_with?("44051401458", :male)
      true

      iex> ExPesel.valid_with?("88122302080", :male)
      false

      iex> ExPesel.valid_with?("44051401458", :female)
      false

      iex> ExPesel.valid_with?("88122302080", :female)
      true

      iex> ExPesel.valid_with?("44051401458", {1944, 5, 14})
      true

      iex> ExPesel.valid_with?("44051401458", {1944, 6, 13})
      false

      iex> ExPesel.valid_with?("some really bad data", {1944, 6, 13})
      false
  """
  def valid_with?(pesel, sex_or_birthday)

  # sorry have to do by def because there is no defdelegate with atoms
  @spec valid_with?(String.t(), :male) :: boolean()
  def valid_with?(pesel, :male), do: Pesel.valid_with?(pesel, :male)

  @spec valid_with?(String.t(), :female) :: boolean()
  def valid_with?(pesel, :female), do: Pesel.valid_with?(pesel, :female)

  @spec valid_with?(String.t(), {1800..2299, 1..12, 1..31}) :: boolean()
  defdelegate valid_with?(pesel, birthdate), to: Pesel

  @doc """
  Returns date of birth extracted from PESEL.

  Date of birth is encoded by first 6 digits as: `yymmdd`

  Century is encoded in `mm` part of it:
  * 1800 - 1899 there is addition of 80 to `mm`
  * 1900 - 1999 there is addition of  0 to `mm`
  * 2000 - 2099 there is addition of 20 to `mm`
  * 2100 - 2199 there is addition of 40 to `mm`
  * 2200 - 2299 there is addition of 50 to `mm`

  For example:
      iex> ExPesel.birthdate("01920300359")
      {1801, 12, 3}

      iex> ExPesel.birthdate("44051401459")
      {1944, 5, 14}

      iex> ExPesel.birthdate("10320305853")
      {2010, 12, 3}

      iex> ExPesel.birthdate("90520308014")
      {2190, 12, 3}

      iex> ExPesel.birthdate("90720312611")
      {2290, 12, 3}

      iex> ExPesel.birthdate("some really bad input")
      :unknown
  """
  @spec birthdate(String.t()) :: {1800..2299, 1..12, 1..31} | :unknown
  defdelegate birthdate(pesel), to: Pesel

  @doc """
  Returns sex extracted from PESEL.

  Sex is encoded by digit at 10 position:
  * even number for female
  * odd number for male

  For example:
      iex> ExPesel.sex("44051401459")
      :male

      iex> ExPesel.sex("88122302080")
      :female

      iex> ExPesel.sex("some bad input")
      :unknown
  """
  @spec sex(String.t()) :: :male | :female | :unknown
  defdelegate sex(pesel), to: Pesel

  @doc """
  Is PESEL number a zombie?

  We got zombie PESEL when it has date of birth 123 years before today.

  For example:
      iex> ExPesel.zombie?("01920300359")
      true

      iex> ExPesel.zombie?("88122302080")
      false

      iex> ExPesel.zombie?("I'm a zombie man!")
      false
  """
  @spec zombie?(String.t()) :: boolean()
  defdelegate zombie?(pesel), to: Pesel
end
